src/box/schema.cc

```diff
@@ -487,3 +489,307 @@ schema_find_name(enum schema_object_type type, uint32_t object_id)
+{
+	char key[16];
+	char *key_end = mp_encode_array(key, 1);
+	key_end = mp_encode_uint(key_end, space_id);
+	struct tuple *tuple;
+	if (box_index_get(BOX_SPACE_ID, 0, key, key_end, &tuple) != 0)
+		return -1;
+	assert(tuple != NULL);
+	const char *data = tuple_data(tuple);
+	const char *data_end = data;
+	mp_next(&data_end);
+
+	const char *str;
+	if (fieldno != UINT32_MAX) {
+		str = tuple_field(tuple, BOX_SPACE_FIELD_FORMAT);
+		assert(str != NULL);
+		mp_decode_array(&str);
+		for (uint32_t i = 0; i < fieldno; ++i)
+			mp_next(&str);
+	} else {
+		str = tuple_field(tuple, BOX_SPACE_FIELD_OPTS);
+		assert(str != NULL);
+	}
+
+	const char *prev = str;
+	uint32_t count = mp_decode_map(&str);
+	const char *mid = str;
+	uint32_t id = 0;
+	const char *constraints = str;
+	const char *constraint_type =
+		type == CONSTR_FKEY ? "foreign_key" : "constraint";
+	for (; id < count; ++id) {
+		uint32_t len;
+		const char *tmp = mp_decode_str(&str, &len);
+		if (len == strlen(constraint_type) &&
+		    memcmp(tmp, constraint_type, len) == 0)
+			break;
+		mp_next(&str);
+		constraints = str;
+	}
+	if (id == count) {
+		int errcode = type == CONSTR_FKEY ? ER_NO_SUCH_FOREIGN_KEY :
+			      ER_NO_SUCH_CONSTRAINT;
+		diag_set(ClientError, errcode, name,
+			 space_name(space_by_id(space_id)));
+		return -1;
+	}
+
+	uint32_t constraint_count = mp_decode_map(&str);
+	const char *constraint_begin = str;
+	const char *constraint = str;
+	uint32_t constraint_id = 0;
+	for (; constraint_id < constraint_count; ++constraint_id) {
+		uint32_t len;
+		const char *tmp = mp_decode_str(&str, &len);
+		if (name_len == len && memcmp(tmp, name, len) == 0)
+			break;
+		mp_next(&str);
+		constraint = str;
+	}
+	if (constraint_id == constraint_count) {
+		int errcode = type == CONSTR_FKEY ? ER_NO_SUCH_FOREIGN_KEY :
+			      ER_NO_SUCH_CONSTRAINT;
+		diag_set(ClientError, errcode, name,
+			 space_name(space_by_id(space_id)));
+		return -1;
+	}
+	const char *post = constraint;
+	mp_next(&post);
+	mp_next(&post);
+	--constraint_count;
+	if (constraint_count == 0)
+		--count;
+
+	uint32_t size = (data_end - data) - (post - constraint);
+	char *new_data = (char *)xmalloc(size);
+	char *new_data_end = new_data;
+	uint32_t prev_size = prev - data;
+	memcpy(new_data_end, data, prev_size);
+	new_data_end += prev_size;
+	new_data_end = mp_encode_map(new_data_end, count);
+	uint32_t mid_size = constraints - mid;
+	memcpy(new_data_end, mid, mid_size);
+	new_data_end += mid_size;
+	if (constraint_count > 0) {
+		new_data_end = mp_encode_str0(new_data_end, constraint_type);
+		new_data_end = mp_encode_map(new_data_end, constraint_count);
+		uint32_t constraint_size = constraint - constraint_begin;
+		memcpy(new_data_end, constraint_begin, constraint_size);
+		new_data_end += constraint_size;
+	}
+	uint32_t post_size = data_end - post;
+	memcpy(new_data_end, post, post_size);
+	new_data_end += post_size;
+	int rc = box_replace(BOX_SPACE_ID, new_data, new_data_end, NULL);
```

Comment:
You might think about `box_update()` with `#` operation. It supports JSON paths, so you can build msgpack operations like `{{'#', 'path.to.constraint.def'}}`. In case it won't fail on a not existing field (I don't remember if it does), then on success you can compare old and new tuples using `memcmp`. If they are equal, it means the update didn't do anything and the constraint wasn't found. `box_update()` could simplify this code a lot.

Even to build the operations msgpack you could use `mp_vformat()` so as not to encode anything manually.

Same for constraints creation.

Reply:
Thank you for the suggestion! The use of JSON paths helped a lot here - now the code is much simpler and a little shorter.